eth,misc: Use transaction manager with confirmation and optional replacement queue

kyriediculous opened the pull request:
**What does this pull request do? Explain your changes. (required)**
Use a `TransactionManager` to handle transactions. This service will forward transactions to the ethereum JSON-RPC, when a transaction is broadcasted to the ethereum network it will be added to the queue. Lowest non-confirmed transactions are popped off the queue to wait until they are mined and retrieve their transaction receipt. 

Once the transaction receipt is found it will be sent over a subscription of transaction receipts that services that want feedback on tx confirmation can subscribe to. This subscription service now powers `LivepeerEthClient.CheckTx()` under the hood. 

**Specific updates (required)**
- 75eed97c84e9d40194403996fc75443f7d375d9e  eth: transaction manager with confirmation queue and optional replacement transactions

-  60d33a0c763505098488050999b220f51ed740a3  pm: handle ticket revert error

- a7b7c5e46e7d406a0f6330f53751a53e20e217ac  eth: remove replacement transactions from reward service

- d4886f688b6edc8f25243d846872a8821ca9b16f  eth,server: move max gas price logic to gas price monitor

- b12a24dc6888ff94d6a68fe3f6628cc828a1e015 eth: make abiMap and txLog helpers in the eth package rather than Backend methods

- 1b89b5c37b1a57da231df387bb8f3d163ed412a0  cmd: replace transaction flag

- efa84b466cebc138501eb8dcb380abe7fbbfd539  cmd,eth: use struct for LivepeerEthClient constructor parameters that aren't other services


**How did you test each of these updates (required)**
- Unit tests

- [ ] TODO add `TransactionManager` unit tests

**Does this pull request close any open issues?**
Fixes #1906

**Checklist:**
- [x] Read the [contribution guide](./doc/contributing.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [x] README and other documentation updated
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

kyriediculous commented on eth/backend.go:
```diff
@@ -42,32 +44,27 @@ type Backend interface {
 	nonceManager *NonceManager
 	signer       types.Signer
+	gpm          *GasPriceMonitor
+	tm           ethereum.TransactionSender
```
prolly better to just refer to `*TransactionManager`

kyriediculous committed:
cmd,eth: use struct for LivepeerEthClient constructor parameters that aren't other services

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -134,6 +134,7 @@ func main() {
 	ethKeystorePath := flag.String("ethKeystorePath", "", "Path for the Eth Key")
 	ethOrchAddr := flag.String("ethOrchAddr", "", "ETH address of an on-chain registered orchestrator")
 	ethUrl := flag.String("ethUrl", "", "Ethereum node JSON-RPC URL")
+	replaceTx := flag.Bool("replaceTransactions", true, "Set to true to replace ethereum transactions with a higher gas price if they time out")
```
What about a `-maxTxReplacements <INTEGER>` flag instead? The value for this flag is the # of times a tx can be replaced (also the # of times the gas price of a tx can be increased). The default value could be something that feels reasonable like 3. The user could set the value to 0 to disable tx replacements.

While an excessively high # of tx replacements may be unlikely in practice if things are implemented correctly, setting a ceiling will be safer to account for pathological cases where the node might just keep endlessly replacing txs. And for users that want to be more careful, they can set a lower ceiling. With this in mind, I think it makes sense to use the same flag to express both enabling/disabling tx replacements and the max # of tx replacements.

kyriediculous replied:
I think this can be a neat addition yeah. One thing that I don't really like here is the hardcoded `EThTxTimeout`, while a user can set the number of max replacements to do it is not immeadiatly known how often the replacement will be tried. 

Allowing a `ethTxTimeout` flag could help here but also again increases the amount of config params required. 

What are your thoughts on that?

yondonfu replied:
Oops missed this comment in the first couple of reviews.

I think it makes sense to allow the timeout to be configurable. What about `-transactionTimeout`?

yondonfu commented on pm/queue.go:
```diff
@@ -131,7 +131,7 @@ ticketLoop:
 						if res.err != nil {
 							glog.Errorf("Error redeeming err=%v", res.err)
-							if res.err != errIsUsedTicket {
+							if res.err != errIsUsedTicket && err != errTicketRevert {
```
```suggestion
							if res.err != errIsUsedTicket && res.err != errTicketRevert {
```

yondonfu commented on eth/backend.go:
```diff
@@ -197,18 +147,48 @@ func (b *backend) retryRemoteCall(remoteCall func() ([]byte, error)) (out []byte
 		parsedAbi, err := abi.JSON(strings.NewReader(ABI))
 		if err != nil {
-			return map[string]*abi.ABI{}, err
+			glog.Errorf("Error creating ABI mappings, can not parse transaction logs err=%v", err)
```
```suggestion
			glog.Errorf("Error creating ABI map err=%v", err)
```

since there is no transaction log parsing in this function.

yondonfu commented on eth/client.go:
```diff
@@ -883,84 +888,31 @@ func (c *client) ContractAddresses() map[string]ethcommon.Address {
+				if event.Status == uint64(0) {
+					return fmt.Errorf("transaction failed txHash=%v", event.TxHash)
+				}
+				break txLoop
```
Can we just return nil here?

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,233 @@
+				}
+				tx = newTx
+			} else {
+				glog.Error(err)
```
There should be an error check here. Otherwise this will log err even if it is nil.

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,233 @@
+
+	tm.feed.Send(receipt)
+
+	if receipt.Status == uint64(0) {
```
What about just letting the caller i.e. CheckTx() handle this check? Otherwise, we currently have this check here and in CheckTx().

yondonfu commented on eth/client.go:
```diff
@@ -883,84 +888,31 @@ func (c *client) ContractAddresses() map[string]ethcommon.Address {
-	if err != nil {
-		return err
+	txs := make(chan *types.Receipt, 10)
+	b, ok := c.backend.(*backend)
```
Could we avoid this type assertion if the TransactionManager was just a field on `client`? Feels a bit weird to do a type assertion just so we can fetch the TransactionManager field on the underlying backend type.

kyriediculous replied:
The contract bindings use `Backend` to send transactions which uses `TransactionManager` under the hood. 

The `Backend` does not have access to the `client`.

yondonfu replied:
I was thinking something like this:

```golang
type client struct {
    // Other fields
    ...
    transactionManager *TransactionManager // This is the same TransactionManager used by the backend
}

func (c *client) CheckTx(tx *types.Transaction) error {
    receipts := make(chan *transactionReceipt, 10)
    txSub := c.transactionManager.Subscribe(receipts)
    defer txSub.Unsubscribe()
    ...
}
```

The main reason for this suggestion is that it feels a bit weird to break the Backend interface abstraction with a type assertion just to fetch the TransactionManager field on the underlying backend type. If we're breaking the Backend interface abstraction within the client then it raises the question of why the Backend interface is being used if we're exposing the implementation details of the underlying type to the client.

yondonfu commented on eth/client.go:
```diff
@@ -883,84 +888,31 @@ func (c *client) ContractAddresses() map[string]ethcommon.Address {
+		case err := <-txSub.Err():
+			return err
+		case event := <-txs:
+			if tx.Hash() == event.TxHash {
```
Looks like this TransactionManager API currently implies that a tx will always be confirmed which is signaled by receiving a receipt in the subscription channel. Do we really want the API to make this guarantee to the caller? If we end up adding a max # of replacements there seems like there could be edge cases where the max # of replacements is hit without the tx being confirmed. In that scenario, a receipt would never be received in the subscription channel and the caller would wait forever. Perhaps the API should give the caller a receipt *or* an error?

kyriediculous replied:
Yeah in retrospect I like your idea better than the current implementation, this will also give us access to revert reason error messages being returned back through `checkTx` 

Updated this API and its implementation in 41d57f11cc1d93e63a2345f082783d00eb600e24

kyriediculous commented:
@yondonfu Do we want persistent storage for the transaction confirmation queue to avoid unexpected behaviour between node restarts when transactions are still pending ?

yondonfu commented on eth/gaspricemonitor.go:
```diff
@@ -125,6 +127,18 @@ func (gpm *GasPriceMonitor) Stop() error {
+	gpm.maxGasPrice = gp
+}
+
+func (gpm *GasPriceMonitor) MaxGasPrice() *big.Int {
```
I think the maxGasPrice field should also be initialized in the constructor.

yondonfu commented on eth/client.go:
```diff
@@ -148,34 +145,42 @@ type client struct {
+	Password          string
+	ControllerAddr    ethcommon.Address
+	TxTimeout         time.Duration
+	MaxGasPrice       *big.Int
```
Doesn't look like this is used anywhere. If we add support for initializing the maxGasPrice field in the GasPriceMonitor constructor we could remove this field since the client is initialized with an existing GasPriceMonitor instance which should already have an initialized maxGasPrice.

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,241 @@
+
+type TransactionManager struct {
+	txTimeout       time.Duration
+	maxReplaceMents int
```
```suggestion
	maxTxReplacements int
```

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,241 @@
+
+type transactionQueue []*types.Transaction
+
+type transactionReceipt struct {
```
Would `transactionResult` be a more clearer name given that this struct already holds a transaction receipt?

yondonfu replied:
Bump.

kyriediculous replied:
I went with `transactionReceipt` since the receipt values are embedded in this type (through albeit through a pointer) with the addition of an error (which can be the revert message in some cases).

yondonfu commented on eth/client.go:
```diff
@@ -148,34 +145,42 @@ type client struct {
-	if err != nil {
-		return nil, err
-	}
+	tm := NewTransactionManager(eth, gpm, am, cfg.TxTimeout, cfg.MaxTxReplacements)
```
Doesn't look like `Start()` and `Stop()` are ever called for the TransactionManager which would be required to start the checkTxLoop.

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,241 @@
+			if receipt, err = tm.wait(tx); err == context.DeadlineExceeded && tm.maxReplaceMents > 0 {
+				newTx, replaceErr := tm.replace(tx)
+				if replaceErr != nil {
+					glog.Error(replaceErr)
```
I think `replaceError` should be propagated back to the feed so that a consumer of the feed can get error information about the failed replacement. At the moment, after we break here, the value sent to the feed will just include the `err` variable which would just be `context.DeadlineExceeded`.

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,241 @@
+				tx = newTx
+			} else {
+				if err != nil {
+					glog.Error(err)
```
I think this error should be propagated back to the feed so that a consumer of the feed can get error information about the failed replacement. At the moment, after we break here, the value sent to the feed will just include the original `err` variable which would be `context.DeadlineExceeded`.

kyriediculous replied:
This error is already propagated back to the feed by default and can only ever result in it being `context.DeadlineExceeded` when the node isn't set to replace transactions in which case we likely want to notify the user to manually check confirmation through this error.

yondonfu replied:
My mistake - I misread the code. You're right that this error is propagated to the feed.

yondonfu commented on pm/sendermonitor.go:
```diff
@@ -422,7 +424,7 @@ func (sm *LocalSenderMonitor) redeemWinningTicket(ticket *SignedTicket) (*types.
 			monitor.TicketRedemptionError()
 		}
-		return nil, err
+		return nil, errTicketRevert
```
Returning `errTicketRevert` here obfuscates the error returned by `CheckTx` which also includes the hash of the reverted tx. As a result, we lose access to that information. Could we continue passing the error from `CheckTx` here and having a different check for a reverted tx error?

yondonfu reviewed (changes requested):
How were these changes tested?

Given that the logic of the transaction manager for handling replacing pending transactions is important for funds management and node operation could we add unit tests particularly around the automated transaction replacement flow?

yondonfu commented on cmd/devtool/devtool.go:
```diff
@@ -159,10 +161,42 @@ func ethSetup(ethAcctAddr, keystoreDir string, isBroadcaster bool) {
+
+	chainID, err := backend.ChainID(context.Background())
+	if err != nil {
+		glog.Errorf("failed to get chain ID from remote ethereum node: %v", err)
```
```suggestion
		glog.Errorf("Failed to get chain ID from remote ethereum node: %v", err)
```

yondonfu commented on cmd/devtool/devtool.go:
```diff
@@ -159,10 +161,42 @@ func ethSetup(ethAcctAddr, keystoreDir string, isBroadcaster bool) {
+
+	am, err := eth.NewAccountManager(ethcommon.HexToAddress(ethAcctAddr), keystoreDir, signer)
+	if err != nil {
+		glog.Errorf("error creating Ethereum account manager: %v", err)
```
```suggestion
		glog.Errorf("Error creating Ethereum account manager: %v", err)
```

yondonfu commented on cmd/devtool/devtool.go:
```diff
@@ -159,10 +161,42 @@ func ethSetup(ethAcctAddr, keystoreDir string, isBroadcaster bool) {
+	}
+
+	tm := eth.NewTransactionManager(backend, gpm, am, 5*time.Minute, 0)
+	tm.Start()
```
This will block so it should be called in a goroutine.

kyriediculous replied:
Sorry missed this one, fixed in the latest commit

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -390,7 +392,41 @@ func main() {
+		// Start gas price monitor
+		_, err = gpm.Start(ctx)
+		if err != nil {
+			glog.Errorf("error starting gas price monitor: %v", err)
```
```suggestion
			glog.Errorf("Error starting gas price monitor: %v", err)
```

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -390,7 +392,41 @@ func main() {
+
+		am, err := eth.NewAccountManager(ethcommon.HexToAddress(*ethAcctAddr), keystoreDir, signer)
+		if err != nil {
+			glog.Errorf("error creating Ethereum account manager: %v", err)
```
```suggestion
			glog.Errorf("Error creating Ethereum account manager: %v", err)
```

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -390,7 +392,41 @@ func main() {
+		}
+
+		if err := am.Unlock(*ethPassword); err != nil {
+			glog.Errorf("error unlocking Ethereum account: %v", err)
```
```suggestion
			glog.Errorf("Error unlocking Ethereum account: %v", err)
```

yondonfu commented on cmd/devtool/devtool.go:
```diff
@@ -159,10 +161,42 @@ func ethSetup(ethAcctAddr, keystoreDir string, isBroadcaster bool) {
+	}
+
+	if err := am.Unlock(passphrase); err != nil {
+		glog.Errorf("error unlocking Ethereum account: %v", err)
```
```suggestion
		glog.Errorf("Error unlocking Ethereum account: %v", err)
```

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -390,7 +392,41 @@ func main() {
+		}
+
+		tm := eth.NewTransactionManager(backend, gpm, am, EthTxTimeout, *maxTxReplacements)
+		tm.Start()
```
This will block so it needs to be called in a goroutine.

yondonfu commented on eth/client.go:
```diff
@@ -883,84 +870,28 @@ func (c *client) ContractAddresses() map[string]ethcommon.Address {
+					return receipt.err
+				}
+				if receipt.Status == uint64(0) {
+					return fmt.Errorf("transaction failed txHash=%v", receipt.TxHash)
```
```suggestion
					return fmt.Errorf("transaction failed txHash=%v", receipt.TxHash.Hex())
```

I think we need to call Hex() in order to log a hex string here?

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,242 @@
+	feed  event.Feed
+	scope event.SubscriptionScope
+
+	eth *ethclient.Client
```
A thought on how we could test the tx replacement behavior:

Replacing this `ethclient.Client` with an interface might make writing tests for the replacement logic easier. 

Something like:

```golang
type TransactionSenderReader interface {
	ethereum.TransactionSender
	ethereum.TransactionReader
	// Required for bind.DeployBackend argument in bind.WaitMined
	CodeAt(context.Context, ethcommon.Address, *big.Int) ([]byte, error)
}
```

Then, a stub implementation of the interface could be used in tests allowing us to control how long it takes for a transaction receipt to be returned in `bind.WaitMined()` since the stub implementation could wait X ms before returning a receipt for a call to `TransactionReceipt()`. If the `txTimeout` is set to Y ms such that Y < X, then we should be able to trigger a timeout which should prompt the TransactionManager to submit a replacement tx. The stub implementation could keep track of the txs sent via `SendTransaction()` and we could check the gas price and nonce for txs to assert that replacement txs are actually being sent. Then, we could also write assertions that the TransactionManager does not submit more than `maxReplacements` replacement txs.

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,242 @@
+			}
+		}
+
+		tm.feed.Send(&transactionReceipt{*receipt, err})
```
This will trigger a nil pointer error if `receipt` is nil which will happen if we encounter an error with the tx such that we get a nil receipt back. If the field on the struct here was a receipt pointer instead of a receipt value we could avoid this and then it would be the responsibility of the feed consumer to only rely on the receipt field if there is a non-nil error.

kyriediculous replied:
Yeah probably better to use a pointer here but copying the actual values felt much cleaner and it does feel like we sometimes might overuse pointers a bit which might lead to memory safety issues that can be hard to debug (started thinking about this after the discussing about golang vs rust in discord). However that's unlikely to be much of an issue here so yeah; pointer.

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,242 @@
+			err     error
+		)
+
+		for i := 0; i <= tm.maxReplacements; i++ {
```
Suggestion for a refactor here that I think flattens the code structure to improve readability:

```golang
receipt, err = tm.wait(tx)
// context.DeadlineExceeded indicates that we hit the txTimeout
// If we hit the txTimeout, replace the tx up to maxReplacements times
for i := 0; err == context.DeadlineExceeded && i < tm.maxReplacements; i++ {
	tx, err = tm.replace(tx)
	// Do not attempt additional replacements if there was an error submitting this
	// replacement tx
	if err != nil {
		break
	}
	receipt, err = tm.wait(tx)
}
```

kyriediculous replied:
I understand that it's a little bit cleaner and shorter but what I generally dislike is having the same function call before the loop (to 'initialize') as inside of it. 

However this suggestion made me realise a small error in the conditional of the loop in my code which also shortens that up.

kyriediculous replied:
Ended up going with the suggestion because otherwise we don't wait for the last iteration of the replacement loop and thus the last replacement tx

yondonfu commented on pm/sendermonitor.go:
```diff
@@ -25,6 +25,8 @@ var unixNow = func() int64 {
 	return time.Now().Unix()
 }
 
+type checkTxError error
```
```suggestion
type errCheckTx error
```

To match the error naming convention used elsewhere in this package.

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -134,6 +135,7 @@ func main() {
 	ethKeystorePath := flag.String("ethKeystorePath", "", "Path for the Eth Key")
 	ethOrchAddr := flag.String("ethOrchAddr", "", "ETH address of an on-chain registered orchestrator")
 	ethUrl := flag.String("ethUrl", "", "Ethereum node JSON-RPC URL")
+	maxTxReplacements := flag.Int("maxTransactionReplacements", 0, "Number of times to automatically replace pending Ethereum transactions")
```
What do you think about setting the default value for this flag to 2? This would enable tx replacements by default which helps address scenarios like super low gas price txs without requiring manual intervention from the node operator. A node operator could always disable the feature by manually passing `-maxTransactionReplacements 0` on startup. The suggestion of 2 is fairly arbitrary although it feels like a reasonable number of times to replace a tx.

kyriediculous replied:
A feature that changes the existing behaviour should in my opinion not be enabled by default.

kyriediculous replied:
But I'm open to discuss this, I feel like if replacements should be the default then I think the default should be one replacement, which at least for the reward calls retains the same behaviour.

yondonfu replied:
> A feature that changes the existing behaviour should in my opinion not be enabled by default.

Well that depends on whether the existing default behavior is a problem. Based on #1906, I think this is a case where the existing default behavior can actually be a problem. Also, if the default is 0, then we'd be changing existing behavior for reward calls except this is actually hampering the node's ability to deal with getting transactions confirmed in a timely manner for reward calls instead of improving it.

> I feel like if replacements should be the default then I think the default should be one replacement

I'd be onboard with a default of 1 to be a bit more conservative with the default.

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,259 @@
+	ctx, cancel := context.WithTimeout(context.Background(), tm.txTimeout)
+	defer cancel()
+
+	receipt, err := bind.WaitMined(ctx, tm.eth, tx)
```
Can shorten this to just:

```golang
return bind.WaitMined(ctx, tm.eth, tx)
```

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,259 @@
+		return nil, ErrReplacingMinedTx
+	}
+
+	// Updated gas price must be at least 10% greater than the gas price used for the original transaction in order
```
Let's move this comment into `calcReplacementGasPrice()`

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,259 @@
+	// Bump gas price exceeds max gas price, return early
+	max := tm.gpm.MaxGasPrice()
+	if gasPrice.Cmp(max) > 0 {
+		return nil, fmt.Errorf("Replacement gas price exceeds max gas price suggested=%v max=%v", gasPrice, max)
```
```suggestion
		return nil, fmt.Errorf("replacement gas price exceeds max gas price suggested=%v max=%v", gasPrice, max)
```

To follow convention of error messages in lowercase (except if its extra annotation text in an error log statement)

yondonfu commented on eth/transactionManager_test.go:
```diff
@@ -0,0 +1,320 @@
+	assert.Nil(tx)
+	assert.EqualError(err, ErrReplacingMinedTx.Error())
+
+	// Test error is eth.TransactionNotFound - fail at next step
```
```suggestion
	// Test error is ethereum.NotFound - fail at next step
```

yondonfu commented on eth/transactionManager_test.go:
```diff
@@ -0,0 +1,320 @@
+
+	// Replacement gas price lower than suggest gas price
+	// Use market gas price
+	gpm.gasPrice = big.NewInt(999)
```
Could be worthwhile to add an assertion here making sure that the GasPriceMonitor gas price is greater than the value returned by `calcReplacementGasPrice(stubTx)`.

yondonfu commented on eth/transactionManager_test.go:
```diff
@@ -0,0 +1,320 @@
+	assert.NotNil(event)
+	assert.Nil(event.err)
+
+	// Wait error no replacements
```
Isn't maxReplacement in the TransactionManager still set to 1 here? And in that case, we get a context.DeadlineExceeded error after 2 calls to wait() with the second being for a replacement tx.

yondonfu commented on eth/transactionManager_test.go:
```diff
@@ -0,0 +1,320 @@
+	eth.err["TransactionByHash"] = errors.New("TransactionByHash error")
+	sink = make(chan *transactionReceipt)
+	sub = tm.Subscribe(sink)
+	time.Sleep(5 * time.Millisecond)
```
Why is there a sleep here after the Subscribe() call, but not after any of the other previous Subscribe() calls?

yondonfu commented on eth/transactionManager_test.go:
```diff
@@ -0,0 +1,320 @@
+	assert.Equal(logsAfter-logsBefore, int64(1))
+}
+
+func TestTransactionManager_CheckTxLoop(t *testing.T) {
```
I think we should add a test to ensure that in the worst case we only submit maxReplacements number of replacement txs. This test would assert that maxReplacements number of replacement txs were submitted, but no more than that.

kyriediculous replied:
Would something like this work 840c9f4 ?

yondonfu replied:
Looks good.

yondonfu commented on eth/transactionManager.go:
```diff
@@ -0,0 +1,259 @@
+	}
+}
+
+func calcReplacementGasPrice(tx *types.Transaction) *big.Int {
```
We don't need to address this in this PR, but just making a note that we will probably want to eventually make the gas price bump a parameter given that its a value that influences how soon a replacement tx will be mined and given that the gas price bump required by an ETH node technically doesn't have to be 10% (that just happens to be the default on geth, but there is an option to customize it).

kyriediculous replied:
Wouldn't that only be supportable by self-hosted geth nodes because if the configuration on the livepeer node is set to a lower replacement bump than i.e. infura you would still see `replacement tx underpriced` no ?

yondonfu replied:
Yeah if your gas price bump is lower than the gas price bump required by the ETH provider you would receive a `replacement tx underpriced` error.

Aside from speeding up the confirmation of replacement txs, the motivation for supporting custom gas price bumps is to give the livepeer node operator to adjust the gas price bump (i.e. increase it) used if the ETH provider requires a gas price bump different (i.e. greater) than the current default of 10%. If the node operator does not control the ETH provider (i.e. self-hosted node(s)) and the ETH provider is configured to use a non-default gas price bump (or if the default ever changes) then the current hardcoded gas price bump in the livepeer node may not work.

Anyways this isn't that crucial at the moment, but just wanted to point this out for future reference.

yondonfu commented on eth/rewardservice.go:
```diff
@@ -91,30 +91,11 @@ func (s *RewardService) tryReward() error {
 		}
 
-		glog.Infof("Called reward for round %v - %v rewards minted", currentRound, FormatUnits(tp.RewardPool, "LPTU"))
+		glog.Infof("Called reward for round %v", currentRound)
```
Just noting that this is related to https://github.com/livepeer/go-livepeer/issues/1822

kyriediculous replied:
Yeah sorry for including it here, it doesn't fully address #1822 in that rather than fixing the value shown I decided to remove it for now to avoid confusion.

yondonfu replied:
Yep that's fine I was just making this note for myself as a reminder.

yondonfu commented on eth/transactionManager_test.go:
```diff
@@ -296,6 +298,7 @@ func TestTransactionManager_CheckTxLoop(t *testing.T) {
 	event := <-sink
 	assert.NotNil(event)
 	assert.Nil(event.err)
+	sub.Unsubscribe()
```
Should we remove the deferred Unsubscribe() call on L295 if we're calling Unsubscribe() for each new subscription at the end of each test case?

yondonfu reviewed (commented):
Just two comments to address:

- [Default flag values](https://github.com/livepeer/go-livepeer/pull/1923#discussion_r669943483)
- [Small test related comment](https://github.com/livepeer/go-livepeer/pull/1923#discussion_r669905869)

Let's also rebase on top of master to resolve the conflicts. Feel free to do this in conjunction with addressing the remaining comments.

kyriediculous committed:
eth: transaction manager implementation

kyriediculous committed:
eth,cmd,server: restructuring in ETH package

kyriediculous committed:
eth,pm: implement new transaction API

kyriediculous committed:
changelog entry

kyriediculous commented:
Rebased, resolved merge conflicts and addresses the remaining changes. 

Also moved min and max gas price completely into `GasPriceMonitor` as discussed previously.

yondonfu commented on eth/backend.go:
```diff
@@ -83,63 +75,43 @@ func (b *backend) PendingNonceAt(ctx context.Context, account common.Address) (u
-	b.RLock()
-	defer b.RUnlock()
-	return b.maxGasPrice
+func (b *backend) GasPriceMonitor() *GasPriceMonitor {
```
Since we're removing the max gas price setters/getters from backend should we also remove the min gas price setters/getters with the expectation that the appropriate setters/getters can be called on the GasPriceMonitor returned here?

kyriediculous replied:
Oh yeah forgot to remove these, the calls were already updated to call them on the GPM instead

kyriediculous replied:
fixed in 91ceac9

yondonfu commented on eth/gaspricemonitor.go:
```diff
@@ -148,6 +151,18 @@ func (gpm *GasPriceMonitor) Stop() error {
 	return nil
 }
 
+func (gpm *GasPriceMonitor) SetMaxGasPrice(gp *big.Int) {
```
This is currently missing the metrics recording call via the monitor package that was included in the [method definition](https://github.com/livepeer/go-livepeer/blob/0a72c0ac52b97b37bd2fb51c865ac3f2a3592bc3/eth/backend.go#L134) in backend.

kyriediculous replied:
fixed in 91ceac9

yondonfu commented on eth/gaspricemonitor.go:
```diff
@@ -54,6 +56,7 @@ func NewGasPriceMonitor(gpo GasPriceOracle, pollingInterval time.Duration, minGa
 		pollingInterval: pollingInterval,
 		gasPrice:        big.NewInt(0),
 		minGasPrice:     minGasP,
+		maxGasPrice:     maxGasPrice,
```
Let's add a metrics recording call for maxGasPrice similar to the one we have for minGasPrice on L52 (will probably need a nil check as well in case maxGasPrice is nil).

kyriediculous replied:
fixed in 91ceac9

yondonfu reviewed (approved):
LGTM after squashing 🚢
